### PR TITLE
[profiled] Config dir change from .profiled to .config/profiled JB#9225

### PR DESCRIPTION
**Pull request by dsuni**
Also adds a oneshot script that moves the old files to the new directory.

### CODE REVIEW

File: database.c [hunk 1]
```diff
@@ -307,17 +307,17 @@ static const char *current_(const char *profile, const char *key)
...
 {
   static char path[256] = "";
 
   if( *path == 0 )
   {
-    snprintf(path, sizeof path, "%s/.profiled", getenv("HOME") ?: "/tmp");
+    if (snprintf(path, sizeof(path), "%s/.config/profiled", getenv("HOME")) < 0)
+    {
+      log_err("Path %s/.config/profiled is too long. Max size %lu\n", getenv("HOME"), sizeof(path));
+      snprintf(path, sizeof(path), "/tmp");
```
Comment (spiiroin) [inline database.c]: This goes all wrong now...

Old code: protected against side effects of NULL $HOME and dontcare if $HOME is bonkers long (under assumption that it will be checked later on or something).

New code: removes NULL protection, adds incorrect "path too long" check (snprintf returns number of chars it would have written if there were space, negative values are something else) and superfluous parentheses for sizeof...

File: rpm/move-profiled-config.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+set -e
+OLD_DIR="$HOME/.profiled"
+NEW_DIR="$HOME/.config/profiled"
+
+mkdir -p "$NEW_DIR"
+if [ -d "$OLD_DIR" ] ; then
+  mv -f "$OLD_DIR/*" "$NEW_DIR"
```
Comment (spiiroin) [inline rpm/move-profiled-config.sh]: When is this executed? Are there any guarantees that profiled is not running? If not, then this is not a hazard free way to do migration. I would have simply loaded from legacy dir and new dir and then removed the legacy dir from the daemon itself -> no chances of timing hazards.

**Review by spiiroin — CHANGES_REQUESTED** [review]

File: database.c (hunk 1 above)
Comment (Thaodan) [reply database.c]: Please don't hardcode $HOME/.config but use $XDG_CONFIG_HOME

File: rpm/move-profiled-config.sh (hunk 2 above)
Comment (Thaodan) [reply rpm/move-profiled-config.sh]: I agree best to do it there.

**Review by Thaodan — CHANGES_REQUESTED** [review]
Please follow the XDG spec properly.

File: database.h [hunk 3]
```diff
@@ -85,6 +85,8 @@ void            database_reload(void);
 
 void            database_set_restart_request_cb(void (*cb)(void));
 
+void            migrate_configs               (const char *legacy, const char *current);
```
Comment (pvuorela) [inline database.h]: No need to expose in the header?

File: database.c [hunk 4]
```diff
@@ -1322,3 +1335,27 @@ database_free_changed_values(profileval_t *values)
...
+{
+  int status;
+  pid_t id = fork();
+  if (id < 0)
+  {
+    log_err("Forking failed with error %d\n", id);
+  }
+  else if (id == 0)
+  {
+    execlp("mv", "mv", "-f", legacy, current, NULL);
```
Comment (pvuorela) [inline database.c]: Should be cleaner moving files in-process instead of forking and executing command line tools. Glib surely has nice options for that.

Besides it's possibly a timing problem if the moving happens externally.

File: database.c [hunk 5]
```diff
@@ -307,17 +308,25 @@ static const char *current_(const char *profile, const char *key)
   return res;
 }
 
-/* ========================================================================= *
- * FIXME: horrible $HOME/.profiled access hacks
- * ========================================================================= */
-
 static const char *datadir(void)
+{
+  const char *confdir = (char*)g_get_user_config_dir();
```
Comment (pvuorela) [inline database.c]: Existing thing, but since playing with glib, could consider fully doing gchar and e.g. g_snprintf(), though that might leak into other parts of the code here.

File: database.c [hunk 6]
```diff
@@ -307,17 +308,25 @@ static const char *current_(const char *profile, const char *key)
...
+}
+
+static const char *legacydir(void)
 {
   static char path[256] = "";
 
   if( *path == 0 )
   {
-    snprintf(path, sizeof path, "%s/.profiled", getenv("HOME") ?: "/tmp");
+    snprintf(path, sizeof(path), "%s/.profiled", getenv("HOME") ?: "/tmp");
```
Comment (Thaodan) [inline database.c]: Please use [GLib.get_home_dir](https://docs.gtk.org/glib/func.get_home_dir.html) instead.

File: database.c [hunk 7]
```diff
@@ -742,6 +751,10 @@ database_init(void)
   custom_work  = xstrfmt("%s.tmp", custom_path);
   custom_back  = xstrfmt("%s.bak", custom_path);
 
+  if (access(current_path, F_OK) != 0)
+  {
```
Comment (Thaodan) [inline database.c:835]: It may make sense to use glibs file_test here instead.

**Review by Thaodan — CHANGES_REQUESTED** [review]
Mostly only what @pvuorela already but also use get_home_dir instead of detecting if `$HOME` is set your self.

I would also prefer if the naming of the old and legacy path is the same all around.

File: database.c (hunk 5 above)
Comment (spiiroin) [reply database.c]: Agree with 'gchar' use - it can be kept internal to the function. And going dynamic would make does-it-fit tests unnecessary too. Something like this perhaps?

    static gchar *path = NULL;
    if( !path )
      path = g_strdup_printf("%s/profiled", g_get_user_config_dir() ?: "/tmp");
    return path;

File: database.c [hunk 8]
```diff
@@ -307,17 +307,26 @@ static const char *current_(const char *profile, const char *key)
...
+  static char path[256] = "";
+
+  if( *path == 0 )
+  {
+    snprintf(path, sizeof(path), "%s/profiled", confdir ?: "/tmp");
+  }
+  return path;
+}
+
+static const char *legacydir(void)
```
Comment (spiiroin) [inline database.c:321]: As the point is "directory that was previously used" -> best fit would be to use the old code as-it-were, just renamed.

File: database.c [hunk 9]
```diff
@@ -728,6 +737,65 @@ database_save_later(int secs)
...
+void
+migrate_configs(const char *legacydir, const char *currentdir)
+{
+  DIR *dptr;
+  struct dirent *dir;
+  FILE *src, *dest;
+  char srcname[260], destname[260];
+  int ch;
+  errno = 0;
+  if (rename(legacydir, currentdir) < 0)
```
Comment (spiiroin) [inline database.c]: I was envisioning literally loading from old place if new place does not exist yet, saving to new place only, and removing old files after successful save i.e. couple of extra lines scattered here and there. But I guess this works too - at the cost of otherwise unnecessary added complexity for dealing with "rename did not work" case below.

File: database.c [hunk 10]
```diff
@@ -728,6 +737,65 @@ database_save_later(int secs)
...
+        if (dir->d_type == DT_REG)
+        {
+          snprintf(srcname, sizeof(srcname), "%s/%s", legacydir, dir->d_name);
+          snprintf(destname, sizeof(destname), "%s/%s", currentdir, dir->d_name);
+          src = fopen(srcname, "r");
+          dest = fopen(destname, "w");
+          if (src == NULL || dest == NULL)
+          {
+            log_err("File open failed: %s %s\n", srcname, destname);
+            return;
```
Comment (spiiroin) [inline database.c]: Case in point: why to avoid midfunction returns? Because they make it easy to leak resources. In this case dir handle and possibly a file handle too. (and rest of the old files are left behind?)

... also, use of copy-a-file helper function would make it easier to follow what is going on.

File: database.c [hunk 11]
```diff
@@ -728,6 +737,65 @@ database_save_later(int secs)
...
+    }
+    dptr = opendir(legacydir);
+    if (dptr)
+    {
+      while ((dir = readdir(dptr)) != NULL)
+      {
+        if (dir->d_type == DT_REG)
+        {
+          snprintf(srcname, sizeof(srcname), "%s/%s", legacydir, dir->d_name);
+          snprintf(destname, sizeof(destname), "%s/%s", currentdir, dir->d_name);
```
Comment (spiiroin) [inline database.c]: In general: 'sizeof' is an operator, not a function -> use "sizeof OBJ", not "sizeof(OBJ)" [and yes, sometimes they are needed and sizeof (TYPE) is different thing etc but still].

File: database.c [hunk 12]
```diff
@@ -728,6 +737,65 @@ database_save_later(int secs)
...
+          src = fopen(srcname, "r");
+          dest = fopen(destname, "w");
+          if (src == NULL || dest == NULL)
+          {
+            log_err("File open failed: %s %s\n", srcname, destname);
+            return;
+          }
+          while((ch = fgetc(src)) != EOF)
+          {
+            fputc(ch, dest);
```
Comment (spiiroin) [inline database.c]: character loop? well, it gets the job done and is needed only once ;-)

File: database.c (hunk 7 above)
Comment (spiiroin) [reply database.c:835]: Possibly, to test for dir not just existence, but the rest of the code uses access() for this stuff

File: database.c [hunk 13]
```diff
@@ -728,6 +737,65 @@ database_save_later(int secs)
...
+          {
+            log_warning("Failed to remove %s with %s\n", srcname, strerror(errno));
+          }
+        }
+      }
+    }
+    else
+    {
+      log_err("Directory open failed: %s\n", legacydir);
+      return;
```
Comment (spiiroin) [inline database.c]: In generic, for readability: Handle errors/unexpected things first, ie. in if-block, not else-block

Also, chances are that directory could be removed even if it can't be read, so... might try it instead of bailing out?

**Review by spiiroin — CHANGES_REQUESTED** [review]

File: database.c (hunk 13 above)
Comment (Thaodan) [reply database.c]: In such cases a goto fail also could work quite nicely.

File: database.h [hunk 14]
```diff
@@ -85,6 +85,7 @@ void            database_reload(void);
 
 void            database_set_restart_request_cb(void (*cb)(void));
 
+
```
Comment (Thaodan) [inline database.h]: Unnecessary whitespace change.

File: database.c (hunk 7 above)
Comment (Thaodan) [reply database.c:835]: Since glib also use access this is more a style thing in this case. I prefer which works better.

File: database.c (hunk 9 above)
Comment (Thaodan) [reply database.c]: I agree better play safe.

File: database.c (hunk 7 above)
Comment (dsuni) [reply database.c:835]: No opinion on access vs. glib. Just thought I should point out that it's not testing for the directory, but actually the main config file ( $HOME/.config/profiled/current ). I just thought it was easier, since that value was already sitting in a variable for me, and functionally it seemed more or less equivalent...

File: database.c [hunk 15]
```diff
@@ -728,6 +735,89 @@ database_save_later(int secs)
...
+}
+
+void
+migrate_configs_hard_way(const char *legacydir, const char *currentdir)
+{
+  DIR *dptr;
+  struct dirent *dir;
+  char srcname[260], destname[260];
+  int result = 0;
+  errno = 0;
```
Comment (spiiroin) [inline database.c]: there ought to be no reason to clear errno -> remove it

File: database.c [hunk 16]
```diff
@@ -728,6 +735,89 @@ database_save_later(int secs)
...
+  if (mkdir(currentdir, 0755) < 0 && errno != EEXIST)
+  {
+    log_err("%s: mkdir: %s\n", currentdir, strerror(errno));
+    return;
+  }
+  dptr = opendir(legacydir);
+  if (!dptr)
+  {
+    log_err("Directory open failed: %s\n", legacydir);
```
Comment (spiiroin) [inline database.c]: opendir sets errno -> log error too

File: database.c (hunk 7 above)
Comment (spiiroin) [reply database.c:835]: IMO access is just fine for the purpose. There is code else where to make sure it is a directory.

But what is missing dealing with (future) clean devices where the directory structure gets created later on when saving data i.e. migration is needed only when new directory does not exist and old directory does exist -> both need to be checked to avoid unnecessary migration attempt that would probably make some noise.

File: database.c [hunk 17]
```diff
@@ -728,6 +735,89 @@ database_save_later(int secs)
   }
 }
 
+int
+copy_file(const char *source, const char *destination)
+{
+  FILE *src, *dest;
+  int ch;
+  src = fopen(source, "r");
+  if (src == NULL)
```
Comment (spiiroin) [inline database.c]: if( src == NULL )

File: database.c [hunk 18]
```diff
@@ -728,6 +735,89 @@ database_save_later(int secs)
...
+  {
+    log_err("File open failed: %s\n", source);
+    return -1;
+  }
+  dest = fopen(destination, "w");
+  if (dest == NULL)
```
Comment (spiiroin) [inline database.c]: if( dest == NULL )

File: database.c [hunk 19]
```diff
@@ -728,6 +735,89 @@ database_save_later(int secs)
...
+  {
+    log_err("File open failed: %s\n", destination);
+    fclose(src);
+    return -1;
+  }
+  while((ch = fgetc(src)) != EOF)
```
Comment (spiiroin) [inline database.c]: while( (ch = fgets(src)) != EOF )

File: database.c [hunk 20]
```diff
@@ -728,6 +735,89 @@ database_save_later(int secs)
...
+
+void
+migrate_configs_hard_way(const char *legacydir, const char *currentdir)
+{
+  DIR *dptr;
+  struct dirent *dir;
+  char srcname[260], destname[260];
+  int result = 0;
+  errno = 0;
+  if (mkdir(currentdir, 0755) < 0 && errno != EEXIST)
```
Comment (spiiroin) [inline database.c]: if( mkdir(currentdir, 0755) < 0 && errno != EEXIST )

File: database.c [hunk 21]
```diff
@@ -728,6 +735,89 @@ database_save_later(int secs)
...
+  {
+    log_err("%s: mkdir: %s\n", currentdir, strerror(errno));
+    return;
+  }
+  dptr = opendir(legacydir);
+  if (!dptr)
```
Comment (spiiroin) [inline database.c]: if( !dptr )

File: database.c [hunk 22]
```diff
@@ -728,6 +735,89 @@ database_save_later(int secs)
...
+  {
+    log_err("Directory open failed: %s\n", legacydir);
+    return;
+  }
+  while ((dir = readdir(dptr)) != NULL)
```
Comment (spiiroin) [inline database.c]: while( (dir = readdir(dptr)) != NULL )

File: database.c [hunk 23]
```diff
@@ -728,6 +735,89 @@ database_save_later(int secs)
...
+  {
+    if (dir->d_type == DT_REG)
```
Comment (spiiroin) [inline database.c]: if( dir->d_type == DT_REG )

File: database.c [hunk 24]
```diff
@@ -728,6 +735,89 @@ database_save_later(int secs)
...
+    {
+      snprintf(srcname, sizeof srcname, "%s/%s", legacydir, dir->d_name);
+      snprintf(destname, sizeof destname, "%s/%s", currentdir, dir->d_name);
+      if (copy_file(srcname, destname) < 0)
```
Comment (spiiroin) [inline database.c]: if( copy_file(srcname, destname) < 0 )

File: database.c [hunk 25]
```diff
@@ -728,6 +735,89 @@ database_save_later(int secs)
...
+      {
+        result = -1;
+        continue;
+      }
+      if (remove(srcname) != 0)
```
Comment (spiiroin) [inline database.c]: if( remove(srcname) != 0 )

File: database.c [hunk 26]
```diff
@@ -728,6 +735,89 @@ database_save_later(int secs)
...
+      {
+        log_warning("Failed to remove %s with %s\n", srcname, strerror(errno));
+      }
+    }
+  }
+  closedir(dptr);
+  if (result < 0 || remove(legacydir) != 0)
```
Comment (spiiroin) [inline database.c]: if( result < 0 || remove(legacydir) != 0 )

File: database.c [hunk 27]
```diff
@@ -728,6 +735,89 @@ database_save_later(int secs)
...
+    log_warning("Failed to remove %s with %s\n", legacydir, strerror(errno));
+  }
+}
+
+
+void
+migrate_configs(const char *legacydir, const char *currentdir)
+{
+  errno = 0;
+  if (rename(legacydir, currentdir) < 0)
```
Comment (spiiroin) [inline database.c]: if( rename(legacydir, currentdir) < 0 )

File: database.c [hunk 28]
```diff
@@ -742,6 +832,10 @@ database_init(void)
   custom_work  = xstrfmt("%s.tmp", custom_path);
   custom_back  = xstrfmt("%s.bak", custom_path);
 
+  if (access(current_path, F_OK) != 0)
```
Comment (spiiroin) [inline database.c]: if( access(current_path, F_OK) != 0 )

File: database.c [hunk 29]
```diff
@@ -728,6 +735,89 @@ database_save_later(int secs)
...
+  {
+    log_warning("Failed to remove %s with %s\n", legacydir, strerror(errno));
+  }
+}
+
+
+void
+migrate_configs(const char *legacydir, const char *currentdir)
+{
+  errno = 0;
```
Comment (spiiroin) [inline database.c]: drop unneeded errno clear

**Review by spiiroin — CHANGES_REQUESTED** [review]
There are comments about style issues and midfunction returns that make my eyes sore but... those do not affect functionality -> feel free to ignore.

But do add that "old dir exists" check as condition before attempting migration.

File: database.c (hunk 9 above)
Comment (spiiroin) [reply database.c]: In retrospect a rename attempts would probably have sufficed. If data is lost, there are fallback values by design. And ambienced is anyway overriding everything. So this is not exactly mission critical and it is kind of better than it needs to be already.

File: database.c [hunk 30]
```diff
@@ -728,6 +735,87 @@ database_save_later(int secs)
   }
 }
 
+int
+copy_file(const char *source, const char *destination)
```
Comment (spiiroin) [inline database.c:740]: all these new functions are used only locally -> might as well be 'static'

File: database.c [hunk 31]
```diff
@@ -728,6 +735,87 @@ database_save_later(int secs)
...
+}
+
+void
+migrate_configs_hard_way(const char *legacydir, const char *currentdir)
+{
+  DIR *dptr;
+  struct dirent *dir;
+  char srcname[260], destname[260];
+  int result = 0;
+  if ( mkdir(currentdir, 0755) < 0 && errno != EEXIST )
```
Comment (spiiroin) [inline database.c]: nit: "if( cond )" not "if ( cond )"

File: database.c [hunk 32]
```diff
@@ -728,6 +735,87 @@ database_save_later(int secs)
...
+  {
+    log_err("%s: mkdir: %s\n", currentdir, strerror(errno));
+    return;
+  }
+  dptr = opendir(legacydir);
+  if ( !dptr )
```
Comment (spiiroin) [inline database.c]: nit: "if( cond )" not "if ( cond )"

File: database.c [hunk 33]
```diff
@@ -728,6 +735,87 @@ database_save_later(int secs)
...
+    }
+  }
+  closedir(dptr);
+  if( result < 0 || remove(legacydir) != 0 )
+  {
+    log_warning("Failed to remove %s with %s\n", legacydir, strerror(errno));
+  }
+}
+
+
```
Comment (spiiroin) [inline database.c]: nit: extra empty line

File: database.c [hunk 34]
```diff
@@ -728,6 +735,87 @@ database_save_later(int secs)
...
+migrate_configs(const char *legacydir, const char *currentdir)
+{
+  if( rename(legacydir, currentdir) < 0 )
+  {
+    log_warning("Rename failed with %s. Falling back to copy/delete.\n", strerror(errno));
+    migrate_configs_hard_way(legacydir, currentdir);
+  }
+}
+
+
```
Comment (spiiroin) [inline database.c]: nit: extra empty line

**Review by spiiroin — APPROVED** [review]
Couple of cosmetic nits, approving already.
